src/lazy/encoder/mod.rs

```diff
@@ -0,0 +1,565 @@
+        todo!()
+    }
+}
+
+/// Incrementally encodes a potentially homogeneously typed Ion list.
+pub struct TextListWriter_1_0<'a, W: Write> {
+    writer: &'a mut LazyRawTextWriter_1_0<W>,
+    depth: usize,
+    // The compiler prevents the list writer from being used after `end()` is called.
+    // However, if the user does not call `end()` and the writer goes out of scope, the output
+    // buffer may be left with invalid Ion. This flag is checked in the `Drop` impl to prevent that
+    // from happening quietly.
+    has_been_closed: bool,
+}
+
+impl<'a, W: Write> TextListWriter_1_0<'a, W> {
+    pub fn new(writer: &'a mut LazyRawTextWriter_1_0<W>, depth: usize) -> IonResult<Self> {
+        let space_after_container_start = writer.whitespace_config.space_after_container_start;
+        write!(writer.output, "[{space_after_container_start}")?;
+        Ok(Self {
+            writer,
+            depth,
+            has_been_closed: false,
+        })
+    }
+}
+
+impl<'a, W: Write> TextListWriter_1_0<'a, W> {
+    /// Writes the provided data as a nested value.
+    fn write<V: WriteAsIon>(&mut self, value: V) -> IonResult<&mut Self> {
+        let indentation = self.whitespace_config().indentation;
+        if !indentation.is_empty() {
+            for _ in 0..self.depth {
+                write!(self.output(), "{indentation}")?;
+            }
+        }
+        value.write_as_ion(self.annotated_value_writer())?;
+        let space_between_nested_values = self.whitespace_config().space_between_nested_values;
+        write!(self.output(), ",{space_between_nested_values}")?;
+        Ok(self)
+    }
+
+    /// Finalizes the list, preventing further values from being written.
+    fn end(mut self) -> IonResult<()> {
+        let space_between_top_level_values =
+            self.whitespace_config().space_between_top_level_values;
+        write!(self.output(), "]{space_between_top_level_values}")?;
+        self.has_been_closed = true;
+        Ok(())
+    }
+
+    fn output(&mut self) -> &mut W {
+        &mut self.writer.output
+    }
+
+    fn whitespace_config(&self) -> &WhitespaceConfig {
+        self.writer.whitespace_config
+    }
+
+    #[inline]
+    fn value_writer(&mut self) -> TextValueWriter_1_0<'_, W> {
+        TextValueWriter_1_0 {
+            writer: self.writer,
+            depth: self.depth,
+        }
+    }
+
+    #[inline]
+    fn annotated_value_writer(&mut self) -> TextAnnotatedValueWriter_1_0<'_, W> {
+        TextAnnotatedValueWriter_1_0 {
+            value_writer: self.value_writer(),
+        }
+    }
+}
+
+impl<'a, W: Write> Drop for TextListWriter_1_0<'a, W> {
+    fn drop(&mut self) {
+        // If the user didn't call `end`, the closing delimiter was not written to output.
+        // It's too late to call it here because we can't return a Result.
+        if !self.has_been_closed {
+            panic!("List writer was dropped without calling `end()`.");
+        }
+    }
+}
```

Comment:
🗺️ This is my compromise in the absence of linear types. We can't `close()` in `Drop` because there's no way for the application to handle a failure there. However, that also means we cannot force the writer to close automatically. This at least makes contract violations noisy -- it's not possible to forget to to close the container and keep going.